File: 1520/1520.py
import sys

sys.setrecursionlimit(10000)

# f = open('input.txt', 'r')
# input = f.readline
input = sys.stdin.readline

MAX = 10001


N,M = map(int, input().split())
board = [[MAX for _ in range(M+2)]]+[[MAX] + list(map(int, input().split())) + [MAX] for _ in range(N)]+[[MAX for _ in range(M+2)]]
visited = [[-1 for _ in range(M+2)] for _ in range(N+2)]

delta = [(-1,0),(1,0),(0,-1),(0,1)]

# visited 체크
# 더 낮은 숫자인지 확인
# 더 낮은 숫자가 아닌 경우 return 0
# 더 낮은 숫자인 경우 visited[y][x] += dfx(ny,nx)
def dfs(y,x):
    # 이미 방문한 경우 처리
    if visited[y][x] != -1:
        return visited[y][x]
    # 종점에 도착한 경우 처리
    if y == N and x == M:
        return 1
    # 처음 도착한 경우 초기화
    visited[y][x] = 0

    for dy,dx in delta:
        ny,nx = y+dy,x+dx
        if board[ny][nx] >= board[y][x]:
            continue
        visited[y][x] += dfs(ny,nx)

    return visited[y][x]

dfs(1,1)

print(visited[1][1])

File: 1520/1520.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var (
	//file, _ = os.Open("input.txt") // 동일 경로의 input.txt 파일을 엽니다.
	//sc      = bufio.NewScanner(file)
	sc      = bufio.NewScanner(os.Stdin)
	wr      = bufio.NewWriter(os.Stdout)
	board   [][]int
	visited [][]int
	N, M    int
	deltaY  = [4]int{0, 0, -1, 1}
	deltaX  = [4]int{-1, 1, 0, 0}
)

func main() {
	input()

	dfs(0, 0)

	fmt.Fprintln(wr, visited[0][0])
	wr.Flush()
}

func dfs(y int, x int) int {
	if visited[y][x] > -1 {
		return visited[y][x]
	}
	if y == N-1 && x == M-1 {
		return 1
	}

	visited[y][x] = 0

	for i := 0; i < 4; i++ {
		ny, nx := y+deltaY[i], x+deltaX[i]

		if ny < 0 || ny >= N || nx < 0 || nx >= M {
			continue
		}
		if board[ny][nx] >= board[y][x] {
			continue
		}

		visited[y][x] += dfs(ny, nx)
	}

	return visited[y][x]
}

func input() {
	// input 을 띄어쓰기, \n 단위로 입력받기
	sc.Split(bufio.ScanWords)

	// 배열에 숫자 입력 받기
	N, M = nextInt(), nextInt()
	board = make([][]int, N)
	visited = make([][]int, N)
	for i := 0; i < N; i++ {
		board[i] = make([]int, M)
		visited[i] = make([]int, M)
		for j := 0; j < M; j++ {
			board[i][j] = nextInt()
			visited[i][j] = -1
		}
	}

}

// 입력 받은 숫자를 Int 로 형변환
func nextInt() int {
	sc.Scan()
	text := sc.Text()
	v, _ := strconv.Atoi(text)
	return v
}
